# The main entry point of workflow.
# After configuring, running snakemake -n in a clone of this repository should successfully execute a dry-run of the workflow.

# The data preprocessing is partially based on the following implementations:
#        * https://github.com/snakemake-workflows/dna-seq-varlociraptor
#        * https://github.com/snakemake-workflows/rna-seq-kallisto-sleuth
#        * https://github.com/snakemake-workflows/chipseq

# The data preprocessing steps include:
#        * simulation of datas for testing was performed with ddRAGE,
#          for more information please see https://ddrage.readthedocs.io/en/latest/getting-started/
#        * raw read quality control with FastQC and MultiQC
#        * adapter trimming with cutadapt

# The data analysis steps with NodeRAD include (noderad.smk):
#        * determine edit distances with minimap2 (rules minimap2_index and minimap2)
#        * construct graph from edit distances with graph-tool (rule noderad_graph)
#        * find representatives solving ILP-problem with pulp (rule noderad_representatives)
#        * find the best solution under consideration of coverage and expected value (rule optimized_solution)


include: "rules/common.smk"
include: "rules/noderad.smk"
include: "rules/qc.smk"
include: "rules/cutadapt.smk"

rule all:
    input: all_input
